workflow-runtime/src/commonMain/kotlin/com/squareup/workflow1/internal/WorkflowNode.kt

```diff
@@ -227,39 +234,59 @@ internal class WorkflowNode<PropsT, StateT, OutputT, RenderingT>(
     workflow: StatefulWorkflow<PropsT, StateT, OutputT, RenderingT>,
     props: PropsT
   ): RenderingT {
-    updateCachedWorkflowInstance(workflow)
-    updatePropsAndState(props)
+    val didUpdateCachedInstance = updatePropsAndState(props, workflow)
 
-    baseRenderContext.unfreeze()
-    val rendering = interceptedWorkflowInstance.render(props, state, context)
-    baseRenderContext.freeze()
+    if (!runtimeConfig.contains(PARTIAL_TREE_RENDERING) ||
+      !lastRendering.isInitialized ||
+      subtreeStateDidChange
+    ) {
+      if (!didUpdateCachedInstance) {
```

Comment:
Would it be simpler to just call `updateCachedWorkflowInstance` unconditionally here? It's idempotent, right?

Reply:
good point, we check equality there, so we could call it again